gcc/rust/util/rust-token-converter.cc

```diff
@@ -0,0 +1,586 @@
+	    ProcMacro::Ident::make_ident (token->as_string ())));
+	  break;
+	// Joint punct
+	case OR:
+	case PIPE_EQ:
+	case CARET_EQ:
+	case RIGHT_SHIFT_EQ:
+	case RIGHT_SHIFT:
+	case GREATER_OR_EQUAL:
+	case MATCH_ARROW:
+	case LESS_OR_EQUAL:
+	case LEFT_SHIFT_EQ:
+	case LEFT_SHIFT:
+	case DIV_EQ:
+	case ELLIPSIS:
+	case DOT_DOT_EQ:
+	case DOT_DOT:
+	case RETURN_TYPE:
+	case MINUS_EQ:
+	case PLUS_EQ:
+	case ASTERISK_EQ:
+	case LOGICAL_AND:
+	case AMP_EQ:
+	case PERCENT_EQ:
+	case SCOPE_RESOLUTION:
+	case NOT_EQUAL:
+	  case EQUAL_EQUAL: {
+	    auto str = token->as_string ();
+	    auto it = str.cbegin ();
+	    for (; it != str.cend () - 1; it++)
+	      {
+		trees.back ().push (ProcMacro::TokenTree::make_tokentree (
+		  ProcMacro::Punct::make_punct (*it, ProcMacro::JOINT)));
+	      }
+	    trees.back ().push (ProcMacro::TokenTree::make_tokentree (
+	      ProcMacro::Punct::make_punct (*it, ProcMacro::ALONE)));
+	  }
+	  break;
+	// Alone punct tokens
+	case EQUAL:
+	case RIGHT_ANGLE:
+	case LEFT_ANGLE:
+	case EXCLAM:
+	case TILDE:
+	case PLUS:
+	case MINUS:
+	case ASTERISK:
+	case DIV:
+	case PERCENT:
+	case CARET:
+	case AMP:
+	case PIPE:
+	case PATTERN_BIND:
+	case DOT:
+	case COMMA:
+	case SEMICOLON:
+	case COLON:
+	case HASH:
+	case DOLLAR_SIGN:
+	case QUESTION_MARK:
+	case SINGLE_QUOTE:
+	  trees.back ().push (ProcMacro::TokenTree::make_tokentree (
+	    ProcMacro::Punct::make_punct (token->as_string ()[0],
+					  ProcMacro::ALONE)));
+	  break;
+	case RIGHT_PAREN:
+	  pop_group (trees, ProcMacro::PARENTHESIS);
+	  break;
+	case RIGHT_CURLY:
+	  pop_group (trees, ProcMacro::BRACE);
+	  break;
+	case RIGHT_SQUARE:
+	  pop_group (trees, ProcMacro::BRACKET);
+	  break;
+	case LEFT_SQUARE:
+	case LEFT_CURLY:
+	case LEFT_PAREN:
+	  trees.push_back (ProcMacro::TokenStream::make_tokenstream ());
+	  break;
+	default:
+	  gcc_unreachable ();
+	}
+    }
+  return trees.back ();
+}
+
+static void
+from_tokenstream (const ProcMacro::TokenStream &ts,
+		  std::vector<const_TokenPtr> &result);
+
+/**
+ * Append the token corresponding to a given Ident to a vector.
+ *
+ * @param literal Reference to the Ident to convert.
+ * @param result Reference to the output vector.
+ */
+static void
+from_ident (const ProcMacro::Ident &ident, std::vector<const_TokenPtr> &result)
+{
+  std::string value (reinterpret_cast<const char *> (ident.val), ident.len);
+  if (ident.is_raw)
+    {
+      value = "r#" + value;
+    }
+
+  // TODO: Inject span -> for now spans are not stored in Ident, once changed
+  // the span should be injected in the built token below.
+  Lexer lexer (value);
+  result.push_back (lexer.peek_token ());
+}
+
+static void
+string_literal (const ProcMacro::StringPayload &payload,
+		std::vector<const_TokenPtr> &result)
+{
+  // TODO: UTF-8 string
+  result.push_back (Token::make_string (
+    Location (),
+    std::string (reinterpret_cast<const char *> (payload.data), payload.len)));
```

Comment:
why is the `reinterpret_cast` needed if `payload` is already `const`?

Reply:
This is due to a rust interface incompatibility, cpp string requires a `const char *` while rust works with `const unsigned char *`. I'm aware this is clunky at best but I have not better solution for now. This will be solved with utf-8 introduction in gccrs though.